File: topos/graphs/uast/mapper_javascript.py
from __future__ import annotations

from tree_sitter import Node

from topos.graphs.uast.mapper_common import map_tree_sitter_to_uast
from topos.graphs.uast.models import UASTNode

_DECLARATION_TYPES = {
    "function_definition": "FunctionDecl",
    "function_declaration": "FunctionDecl",  # JS native name for top-level fn
    "function": "FunctionDecl",
    "function_expression": "FunctionDecl",
    "arrow_function": "FunctionDecl",
    "class_definition": "TypeDecl",
    "class_declaration": "TypeDecl",
    "struct_item": "TypeDecl",
    "enum_item": "TypeDecl",
    "impl_item": "TypeDecl",
    "function_item": "FunctionDecl",
    "method_definition": "MethodDecl",
    "lexical_declaration": "VarDecl",
    "variable_declaration": "VarDecl",
}

_STATEMENT_TYPES = {
    "if_statement": "IfStmt",
    "for_statement": "ForStmt",
    "while_statement": "WhileStmt",
    "match_statement": "MatchStmt",
    "return_statement": "ReturnStmt",
    "break_statement": "BreakStmt",
    "continue_statement": "ContinueStmt",
    "throw_statement": "ThrowStmt",
    "try_statement": "TryStmt",
    "expression_statement": "ExprStmt",
}

_EXPRESSION_TYPES = {
    "assignment": "AssignExpr",
    "augmented_assignment": "AssignExpr",
    "binary_expression": "BinaryExpr",
    "boolean_operator": "BinaryExpr",
    "unary_expression": "UnaryExpr",
    "call_expression": "CallExpr",
    "member_expression": "MemberExpr",
    "field_expression": "MemberExpr",
    "subscript": "MemberExpr",
}


def map_node_kind(node: Node) -> str:
    if node.type in _DECLARATION_TYPES:
        return _DECLARATION_TYPES[node.type]
    if node.type in _STATEMENT_TYPES:
        return _STATEMENT_TYPES[node.type]
    if node.type in _EXPRESSION_TYPES:
        return _EXPRESSION_TYPES[node.type]
    if node.type == "identifier":
        return "Identifier"
    if node.type.endswith("literal") or node.type in {"string", "integer", "float"}:
        return "Literal"
    if node.type in {"module", "program", "translation_unit", "source_file"}:
        return "File"
    return "Unknown"


def map_javascript_tree_to_uast(root: Node, file: str | None = None) -> UASTNode:
    # Deliberately no `extract_attributes`/`typeKind` hook here, unlike the
    # Rust/Python/Go/TypeScript mappers: plain JS has no `interface` or
    # `abstract class` syntax at all (that's a TypeScript-only extension of
    # this shared grammar — see mapper_typescript.py), so there's nothing
    # in a .js file's grammar to ever classify as abstract vs. concrete.
    # Martin's Abstractness metric (mdg.abstractness, issue #124) is
    # therefore not meaningful for JavaScript; see the language allowlist
    # in topos/graphs/uast/object.py.
    return map_tree_sitter_to_uast(
        root=root,
        language="javascript",
        map_node_kind=map_node_kind,
        file=file,
    )

File: topos/engine/src/graphs/uast/mapper_javascript.rs
//! JavaScript → UAST mapper.

use std::collections::HashMap;

use tree_sitter::Node;

use super::mapper_common::{logical_operator_attribute, map_tree_sitter_to_uast};
use super::models::{AttributeValue, UASTNode};

pub fn map_node_kind(kind: &str) -> &'static str {
    const NODE_KIND_TABLE: &[(&str, &str)] = &[
        ("function_definition", "FunctionDecl"),
        ("function_declaration", "FunctionDecl"),
        ("function", "FunctionDecl"),
        ("function_expression", "FunctionDecl"),
        ("arrow_function", "FunctionDecl"),
        ("class_definition", "TypeDecl"),
        ("class_declaration", "TypeDecl"),
        ("method_definition", "MethodDecl"),
        ("lexical_declaration", "VarDecl"),
        ("variable_declaration", "VarDecl"),
        ("if_statement", "IfStmt"),
        ("for_statement", "ForStmt"),
        ("while_statement", "WhileStmt"),
        ("switch_statement", "MatchStmt"),
        ("return_statement", "ReturnStmt"),
        ("break_statement", "BreakStmt"),
        ("continue_statement", "ContinueStmt"),
        ("throw_statement", "ThrowStmt"),
        ("try_statement", "TryStmt"),
        ("catch_clause", "CatchClause"),
        ("expression_statement", "ExprStmt"),
        ("assignment", "AssignExpr"),
        ("augmented_assignment", "AssignExpr"),
        ("binary_expression", "BinaryExpr"),
        ("boolean_operator", "BinaryExpr"),
        ("ternary_expression", "ConditionalExpr"),
        ("unary_expression", "UnaryExpr"),
        ("call_expression", "CallExpr"),
        ("member_expression", "MemberExpr"),
        ("field_expression", "MemberExpr"),
        ("subscript", "MemberExpr"),
        ("identifier", "Identifier"),
        ("module", "File"),
        ("program", "File"),
        ("translation_unit", "File"),
        ("source_file", "File"),
    ];
    if let Some((_, mapped)) = NODE_KIND_TABLE.iter().find(|(k, _)| *k == kind) {
        return mapped;
    }
    if kind.ends_with("literal") || matches!(kind, "string" | "integer" | "float") {
        return "Literal";
    }
    "Unknown"
}

fn extract_attributes(node: &Node, source: &[u8]) -> HashMap<String, AttributeValue> {
    logical_operator_attribute(node, source)
}

pub fn map_javascript_tree_to_uast(root: Node, source: &[u8], file: Option<&str>) -> UASTNode {
    map_tree_sitter_to_uast(
        root,
        "javascript",
        map_node_kind,
        source,
        file,
        None,
        Some(&extract_attributes),
    )
}

#[cfg(test)]
mod tests {
    use super::super::models::AttributeValue;
    use super::*;
    use tree_sitter::Parser;

    fn parse(source: &str) -> tree_sitter::Tree {
        let mut parser = Parser::new();
        parser
            .set_language(&tree_sitter_javascript::LANGUAGE.into())
            .unwrap();
        parser.parse(source, None).unwrap()
    }

    fn collect_binary_operators(node: &UASTNode, out: &mut Vec<String>) {
        if node.kind == "BinaryExpr" {
            if let Some(AttributeValue::Str(op)) = node.attributes.get("operator") {
                out.push(op.clone());
            }
        }
        for child in &node.children {
            collect_binary_operators(child, out);
        }
    }

    #[test]
    fn records_logical_operators_on_binary_expr() {
        let source = "function f(a, b, c) { return a && b && c; }\n";
        let tree = parse(source);
        let uast = map_javascript_tree_to_uast(tree.root_node(), source.as_bytes(), None);
        let mut ops = Vec::new();
        collect_binary_operators(&uast, &mut ops);
        assert_eq!(ops, vec!["&&", "&&"]);
    }
}
